### ISSUE TITLE
[Ingest Pipeline] Ability to split documents

### ISSUE THREAD

**Issue by m-adams** [elastic/elasticsearch#56769]
It is common for tools to output data in a combined format where one document may contain several entities.
For example, a tool that scans several hosts for compliance or vulnerabilities or an API that provides an update to every train/bus etc.
We really want to split all entities out to separate docs while copying some high-level information.
This is possible using Logstash and the Split filter but not possible with Ingest Pipelines.

The feature would allow this kind of document to be processed and split without having to include Logstash in the ingest chain.

**Comment by ianmuscat** [comment]
If it helps, this was my use case for needing this feature https://discuss.elastic.co/t/split-json-array-into-multiple-events-using-ingest-pipelines/238519

**Comment by jeffvestal** [comment]
This would be really helpful. 
I'm sending in batches of metrics from a source where it is difficult to change the format to have them sent in one at a time or in bulk format. If I could split the array of json measurements up with an ingest pipeline I could avoid having to put in an additional parsing layer.  

example part of an ingested doc. I want to split out the json measurements in `data` into individual documents
```
 {
        "_index" : "metrics_test",
        "_type" : "_doc",
        "_id" : "IsWOhHcBCOjOntNtCq76",
        "_score" : 1.0,
        "_source" : {
          "data" : [
            {
              "type" : "measure1",
              "date" : "2021-02-08T00:44:32-06:00",
              "value" : "164",
              "unit" : "count"
            },
            {
              "type" : "measure1",
              "date" : "2021-02-08T00:55:16-06:00",
              "value" : "22",
              "unit" : "count"
            },
...
```

**Comment by kofi-grid** [comment]
I could also use this. I'm trying to cut out logstash and can't without this!

**Comment by hungnguyen-elastic** [comment]
this is very much needed! +1

**Comment by quangthien1803** [comment]
this is very much needed! +1

**Comment by ChenTsungYu** [comment]
this is very much needed! +1

**Comment by christophercutajar** [comment]
this is very much needed! +1

**Comment by cvanhalt** [comment]
this is very much needed! +1

**Comment by SpencerLN** [comment]
This is one of the final remaining items preventing us from decommissioning our Logstash instances and fully migrating to beats + Ingest Pipeline. We have multiple data sources that include arrays in the JSON data that need to be split into their own documents while potentially inheriting some properties from the parent document.

For example, input:
```
{
    "user_id": "abc123",
    "time": "1994-11-05T13:15:30Z",
    "events": [
        {
            "event_name": "view_page",
            "event_metadata": "blah"
        },
        {
            "event_name": "click_submit",
            "event_metadata": "blah"
        }
    ]
}
```

output:
```
{
    "user_id": "abc123",
    "time": "1994-11-05T13:15:30Z",
    "event": {
        "event_name": "view_page",
        "event_metadata": "blah"
    }
}
{
    "user_id": "abc123",
    "time": "1994-11-05T13:15:30Z",
    "event": {
        "event_name": "click_submit",
        "event_metadata": "blah"
    }
}
```

**Comment by burgermannetje** [comment]
Is there any confirmation that this will be realized? I can see this has been added to the enhancements and 'needs-triage'. Is there any information on where this is in the works?

**Comment by hendry-lim** [comment]
+1
We have the same requirements from our customer.
We are enriching our data with the `enrich` processor that may match multiple documents.
We need to index the same document + enriched fields into multiple documents if the enrichment matched multiple documents.

**Comment by LeonBirk** [comment]
We have the same issue - the split processor httpjson inputof the filebeat threatintel module does not work properly for our use case (get Attributes from events in MISP as documents). 

```
response.split:
  target: body.response
  split:
    target: body.Event.Object
    split:
      target: body.Event.Object.Attribute
```
still leaves us with 

```
    {
        "Attribute": [
            {
                "category": "Network activity",
                "deleted": false,
                "to_ids": true,
                "value": "https://redacted.net/ls/click?upn=5c-2BN7OI7J"
            },
            {
                "category": "Network activity",
                "to_ids": true,
                "type": "domain",
                "uuid": "76bfee8d-4d2f-4aee-aba6-ab714b1e65ab",
                "value": "redacted.net"
            }
        ],
        "ObjectReference": [
            {
                "Object": {
                    "distribution": "5"
                },
                "comment": "",
                "deleted": false,
                "uuid": "895b6048-1bb1-4f6a-bdb4-cf7fb45f4fcc"
            }
        ],
        "comment": "Redirector URL contained in mail",
        "event_id": "3835"

    }
```

This could be resolved through splitting docs with an ingest pipeline.

**Comment by nl19856** [comment]
I am also eager to have it.

**Comment by emmanuelmathot** [comment]
+1

**Comment by legoguy1000** [comment]
I too would love to see this feature

**Comment by smnschndr** [comment]
+1

**Comment by felix-lessoer** [comment]
Pretty important feature!

**Comment by miastensdotter** [comment]
+1

**Comment by itaykat** [comment]
+1

**Comment by luizhlelis** [comment]
+1

**Comment by ylasri** [comment]
+1

**Comment by renormalist** [comment]
+1

**Comment by JeppeMariagerLam** [comment]
+1

**Comment by Rick25-dev** [comment]
Any update? It has been 2 years @elastic/es-core-features (:Core/Features/Ingest)

**Comment by praveenvinay** [comment]
+1

**Comment by ericleong86** [comment]
+1 this is needed. We often query APIs to multiple products and it always return results in an array format. Need to split the results into multiple documents instead. Creating different API requests to call for specific object is not practical in my use case.

**Comment by fatihakafou** [comment]
+1 This is needed!

**Comment by LBoraz** [comment]
+1 lack of this feature forces an unnecessary trip to logstash

**Comment by djptek** [comment]
Depending on your use use case, Filebeat has a processor which 

https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-httpjson.html#response-split

which may be relevant as it allows you to:

"...convert a map, array, or string into multiple events..."

and optionally

"...fields from the parent document (at the same level as target) will be kept..."

**Comment by expertalex** [comment]
+1 it will help us avoid using [elastic-serverless-forwarder ](https://github.com/elastic/elastic-serverless-forwarder/blob/main/docs/README-AWS.md#expanding-events-from-list-in-json-object) with expand_event_list_from_field

**Comment by janniten** [comment]
+ 1 needed for vulnerability/Missing Kbs Ingestion

**Comment by fixcer** [comment]
+1 Pretty important feature!

**Comment by skbriink** [comment]
+1 would love this!

**Comment by tomgregoryelastic** [comment]
+1

**Comment by timor-raiman** [comment]
+1

**Comment by jeskoriesner** [comment]
+1

**Comment by renzedj** [comment]
+1 - this just cropped up for me. It's the absolute ONLY thing that would (almost) force me to bring back Logstash into my environment.

Blech.

**Comment by shiya-kohn** [comment]
👍 We need this for blue-green sharding

**Comment by llermaly** [comment]
This would be very useful to split text body into chunks to overcome the 512 tokens limit of the embeddings models

**Comment by luc-anise** [comment]
+1

**Comment by carlopuri** [comment]
Please add this feature; I'll solve use cases with feature enabled on ingest pipeline. Thanks a lot

**Comment by DrMxxxxx** [comment]
This would be a veeeeery helpful feature. I have the need to split an array in one document to several documents.

**Comment by thekofimensah** [comment]
It's been 3 years @elastic How many more +1s are needed?

**Comment by dijkstrajesse** [comment]
+1

**Comment by Lexinga** [comment]
+1

**Comment by MichelKoba** [comment]
+1

**Comment by behnam-ramouzeh** [comment]
+1

**Comment by birkanatici** [comment]
➕1️⃣

**Comment by dstor-team** [comment]
+1

**Comment by clement-fouque** [comment]
This would be extremely helpfull and would allow us to solve this issue: https://github.com/elastic/integrations/issues/9261.

**Comment by jmbass** [comment]
+1

**Comment by tylerperk** [comment]
Indeed this is a popular request and we recognize there are valid use cases for this. However I think we are unlikely to implement this anytime soon given the complexity involved and balancing investment against our other priorities. We have discussed it on occasion within the team and it is viewed as extremely complex to implement and fraught with edge cases that we would want to cover if it were to be a robust and reliable solution. 

Some of the concerns raised are about how to provide effective failure handling behavior and response details: 
- Unclear how to handle scenarios where you send a document that needs to be created and it splits into two documents, and one document is accepted and the other document is not
- What to tell the requestor and how it should react 
- What to do with the document that was created: We wouldn’t be able to rollback the previous document without a follow up deletion, which also could fail for any reason. If we leave the first document in and the client retries sending the original create request to create the second one, we will have a case where the first subdocument is doomed to fail because it already exists
_- Would you be willing to accept compromised failure handling as a tradeoff for the ability to split documents - would a very basic implementation be acceptable?_

As well as bookkeeping for bulk requests:
- Reporting results in the response becomes tricky as well since each incoming document that is split will need its position in the bulk request updated with a sub-response for each split-off document. If you can split documents off of a regular document, is it possible that those could split further still? 
- Additionally, a lot of our bulk logic operates on batches of documents that are arranged in the same order that they arrived from the bulk request. Generating new documents will need additional bookkeeping in the bulk code since all documents will now need to be remapped back to their original place in the first bulk request. We do this in some places, but it’s always a subdivision of the incoming data which allows for stable mappings in their positions. Adding new documents after a request arrives throws that out of whack

Workaround options include using Logstash or other external tools to do the splitting. Doing it outside of ES helps with many of the concerns we have about doing it in ES. As we move more towards embracing OTel, in the future you might be able to use an OTel collector to do the split or it might be possible in other parts of Elastic offerings (no promises, just thinking out loud there). 

In summary, we have no plans to implement this but please feel free to add more use case examples which would be valuable input to requirements should we choose to do this later.

**Comment by carlopuri** [comment]
Hello,

we have 2 use cases that can be solved only with this ingest pipeline enhancement.

**Use case 1 - "inference" processor with "completion" task to prompt OpenAI and then manage GPT answer**
Whit this implementation, during document ingestion, we prompt OpenAI asking to analyse the text and give us back the result in a JSON document, with some fields as array.
To use the data in Kibana we need to flatten all the received arrays.

Below an example of a document to index:
```
{
     "_index": "test-inference-ai-001",
     "_id": "test_ai_000001",
     "text_to_analyze": "Can't use the app or website to top up with my valid credit and debit cards. On the app it says declied by bank but on the website, it pulls the funds from my acct, then reverses the transaction 30 minutes later. A total waste of time. Contacted the bank and they confirmed they aren't declining it and the customer service chat was useless."
}
```

and the retuned JSON from OpenAI through inference processor with completion task:
```
{
  "topics": ["app", "website", "customer service"],
  "sentiments": ["negative", "negative", "negative"],
  "segments": ["Can't use the app or website to top up with my valid credit and debit cards.", "On the app it says declined by bank but on the website, it pulls the funds from my acct, then reverses the transaction 30 minutes later.", "the customer service chat was useless."],
  "reasons": ["can't top up with valid cards", "pulls funds then reverses transaction", "useless"]
}
```


we need to flatten the arrays **in a new index** like below:

```
{
    "index": "test-inference-ai-001_ingest",
    "_id" : "test_ai_000001_1",
    "topics": "app",
    "sentiments": "negative",
    "segments": "Can't use the app or website to top up with my valid credit and debit cards."
},
{
    "index": "test-inference-ai-001_ingest",
    "_id" : "test_ai_000001_2",
    "topics": "website",
    "sentiments": "negative",
    "segments": "On the app it says declined by bank but on the website, it pulls the funds from my acct, then reverses the transaction 30 minutes later."
},
{
    "index": "test-inference-ai-001_ingest",
    "_id" : "test_ai_000001_3",
    "topics": "customer service",
    "sentiments": "negative",
    "segments": "the customer service chat was useless."
}
```


**Use case 2 - flattening array with "main category" and "sub category" labelling**
Whit this implementation, during document ingestion, we get documents with "main category" and "sub category" arrays.
Both can have multiple values and to use the data in Kibana we need to flatten all these two arrays.
We can't use logstash in our environment.

Below an example of a document to index:
 ```
   {
        "_index": "test-multilabel",
        "_id": "record_001",
        "_score": 1,
        "_source": {
          "doc_created_time": "01/01/2024",
          "top_category_classification": [
            "communications",
            "operator"
          ],
          "subcategory_classification": [
            "communications_general",
            "communications_clear",
            "operator_firendly",
            "operator_fast"
          ],
          "score": 4,
          "segment": "detractors",
          "system_virtual_id": "100200300400500"
    }
```

we need to flatten the arrays **in a new index** like below:

```
{
        "_index": "test-multilabel_splitted",
        "_id": "record_001_1",
        "_score": 1,
        "_source": {
          "doc_created_time": "01/01/2024",
          "top_category_classification": "communications",
          "subcategory_classification": "general",
          "score": 4,
          "segment": "detractors",
          "system_virtual_id": "100200300400500"
},
{
        "_index": "test-multilabel_splitted",
        "_id": "record_001_2",
        "_score": 1,
        "_source": {
          "doc_created_time": "01/01/2024",
          "top_category_classification": "communications",
          "subcategory_classification": "clear",
          "score": 4,
          "segment": "detractors",
          "system_virtual_id": "100200300400500"
},
{
        "_index": "test-multilabel_splitted",
        "_id": "record_001_3",
        "_score": 1,
        "_source": {
          "doc_created_time": "01/01/2024",
          "top_category_classification": "operator",
          "subcategory_classification": "friendly",
          "score": 4,
          "segment": "detractors",
          "system_virtual_id": "100200300400500"
},
{
        "_index": "test-multilabel_splitted",
        "_id": "record_001_4",
        "_score": 1,
        "_source": {
          "doc_created_time": "01/01/2024",
          "top_category_classification": "operator",
          "subcategory_classification": "fast",
          "score": 4,
          "segment": "detractors",
          "system_virtual_id": "100200300400500"
}
```
Using script processor we are able to properly create the documents as a new array (called "items") and we tried to use "foreach" processor on items of that array, but we aren't able to ingest them as separated documents.
Something like:
```
"foreach": {
        "field": "items",
        "processor": {
          "script": {
            "source": """
              if (ctx.items instanceof List) {
                for (int i = 0; i < ctx.items.size(); i++) {
                  def item = ctx.items.get(i);
                  def new_doc = new HashMap(ctx);
                  new_doc.remove('items');
                  new_doc.put('array_item', item);
                  ctx['_ingest']['_index'] = ctx['_index'] + '_splitted';
                  ctx['_ingest']['_id'] = ctx['_id'] + '_' + i;
                  ctx['_ingest']['_op_type'] = 'create';
                  ctx['_ingest']['_bulk'].add(new_doc);
                }
                ctx.op = 'delete';
              }
            """
          }
        }
      }
```

Both of use cases have hight impact in projects with our customers.

In the first use case we are able to do an impressive work with "inference" processor and "completion" task prompting OpenAI but we can't manage and process received data as we need to show results in Kibana.

We hope the feature in this thread will be released soon...

**Comment by syepes** [comment]
+1

**Comment by bfmags** [comment]
+1

**Comment by defutek-tj** [comment]
+1

**Comment by wandergeek** [comment]
+1

**Comment by weibeu** [comment]
+1

**Comment by weibeu** [comment]
+1
Our usecase is that we want to be able to transform the operations from each of the AI agents made to the more ui friendly / query-able format in flattened structure while also preserving the original raw data from the agents intact to feed it back into them as agentic operations / session resumability.

**Comment by Taity045** [comment]
+1 but I guess not getting this anytime soon
